jetty-io/src/main/java/org/eclipse/jetty/io/WriteFlusher.java

```diff
@@ -312,6 +258,8 @@ public InvocationType getCallbackInvocationType()
      */
     public void write(Callback callback, ByteBuffer... buffers) throws WritePendingException
     {
+        callback = Objects.requireNonNull(callback);
+
         if (DEBUG)
             LOG.debug("write: {} {}", this, BufferUtil.toDetailString(buffers));
 
```

Comment:
Throwing the WritePendingException below is now not always true.     If the state is failed, then we should probably throw a new EoFException with a cause from the failed state.

Reply:
@gregw getting a lot of problems adding and `EoFException` here
have to add EofException to the signature then theres all these problems with other methods which use it like AbstractEndPoint and the signatures of those would have to be changed

Reply:
@lachlan-roberts we should not throw, we should fail the callback.

Reply:
I agree with Simone that for Failed state we should call the callback onFail with the cause and then return.

However, I'm conflicted about the WritePendingException, as that is essentially an IllegalStateException and means the API has been used incorrectly, so I think in that case it is OK to continue throwing it.  @sbordet do you agree?

Reply:
@gregw I agree on both (thanks for clarifying). Fail the callback for `FAILED` state and throw `WritePendingException` for incorrect API usage (which would be a Jetty bug as Jetty itself only should call this API).